[ews-build.webkit.org] Install pre-push hook

JonWBedard opened the pull request:
#### 75336e867e1978743991b08f7df8b5a7f1c01626
<pre>
[ews-build.webkit.org] Install pre-push hook
<a href="https://bugs.webkit.org/show_bug.cgi?id=254167">https://bugs.webkit.org/show_bug.cgi?id=254167</a>
rdar://106954459

Reviewed by Aakash Jain.

The pre-push hook in the WebKit repository prevents users
from inadvertently publishing content from a secure remote.
EWS (merge-queue, in particular) should receive the same
protection.

* Tools/CISupport/ews-build/factories.py:
(Factory.__init__): Add InstallHooks step.
(StyleFactory.__init__): Ditto.
(WatchListFactory.__init__): Ditto.
(CommitQueueFactory.__init__): Ditto.
(MergeQueueFactoryBase.__init__): Ditto.
* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps): Add install-hooks step.
* Tools/CISupport/ews-build/steps.py:
(InstallHooks.__init__):
(InstallHooks.run): Add explicit &apos;level&apos; option to
enable pushing the current change to the target remote.
(InstallHooks.getResultSummary):
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/261923@main">https://commits.webkit.org/261923@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d14bd89777ab42b2ec07880f289ff9203ece44f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/6/builds/113292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/22425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/90/builds/1960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/121745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/119078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/90/builds/1960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/106388 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/90/builds/1960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/106388 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/14725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/90/builds/1960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/106388 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/15433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/20/builds/112056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/20727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/90/builds/1960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/17288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -915,6 +915,40 @@ def hideStepIf(self, results, step):
+
+    def run(self):
+        commands = [
+            ['git', 'config', 'include.path', '../metadata/git_config_extension'], [
```
Nit: please verify formatting

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -915,6 +915,40 @@ def hideStepIf(self, results, step):
+        source = self.getProperty('github.head.repo.full_name', None)
+        project = self.getProperty('project', None)
+        if project in GITHUB_PROJECTS and source:
+            commands[-1] += ['--level', 'github.com:{}={}'.format(source, GITHUB_PROJECTS.index(project))]
```
modifying last element (commands[-1]) seems error-prone as someone else can modify the commands list later on without realizing that last element is special.

JonWBedard replied:
Good point, making variable names more explicit

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -915,6 +915,40 @@ def hideStepIf(self, results, step):
+
+    def run(self):
+        commands = [
+            ['git', 'config', 'include.path', '../metadata/git_config_extension'], [
```
is ".." in "../metadata/git_config_extension" expected?

JonWBedard replied:
Yes! It's got to be relative to `<repo>/.git`, and our `git_config_extension` is in `<repo>/metadata/git_config_extension`

aj062 reviewed (approved):
r=me assuming this is tested.

JonWBedard committed:
[ews-build.webkit.org] Install pre-push hook
https://bugs.webkit.org/show_bug.cgi?id=254167
rdar://106954459

Reviewed by Aakash Jain.

The pre-push hook in the WebKit repository prevents users
from inadvertently publishing content from a secure remote.
EWS (merge-queue, in particular) should receive the same
protection.

* Tools/CISupport/ews-build/factories.py:
(Factory.__init__): Add InstallHooks step.
(StyleFactory.__init__): Ditto.
(WatchListFactory.__init__): Ditto.
(CommitQueueFactory.__init__): Ditto.
(MergeQueueFactoryBase.__init__): Ditto.
* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps): Add install-hooks step.
* Tools/CISupport/ews-build/steps.py:
(InstallHooks.__init__):
(InstallHooks.run): Add explicit 'level' option to
enable pushing the current change to the target remote.
(InstallHooks.getResultSummary):
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/261923@main
